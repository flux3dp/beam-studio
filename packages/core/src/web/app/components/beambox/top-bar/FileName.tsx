import React, { useEffect } from 'react';
import ReactDomServer from 'react-dom/server';

import currentFileManager from 'app/svgedit/currentFileManager';
import TopBarController from 'app/views/beambox/TopBar/contexts/TopBarController';
import TopBarIcons from 'app/icons/top-bar/TopBarIcons';
import useForceUpdate from 'helpers/use-force-update';
import useI18n from 'helpers/useI18n';

import styles from './FileName.module.scss';

interface Props {
  hasUnsavedChange: boolean;
  isTitle?: boolean;
}

function FileName({ hasUnsavedChange, isTitle = false }: Props): JSX.Element {
  const forceUpdate = useForceUpdate();
  useEffect(() => {
    TopBarController.onTitleChange(forceUpdate);
    return () => {
      TopBarController.offTitleChange(forceUpdate);
    };
  }, [forceUpdate]);
  const lang = useI18n().topbar;
  const { isCloudFile } = currentFileManager;
  const fileName = currentFileManager.getName() || lang.untitled;
  return (
    <div className={isTitle ? styles.title : styles['file-name']}>
      {isCloudFile && <TopBarIcons.CloudFile className={styles.cloud} />}
      {`${fileName}${hasUnsavedChange ? '*' : ''}`}
    </div>
  );
}

const updateTitle = () => {
  if (window.os === 'Windows' && window.titlebar) {
    const title = ReactDomServer.renderToStaticMarkup(
      <FileName hasUnsavedChange={false} isTitle />
    );
    // eslint-disable-next-line no-underscore-dangle
    if (window.titlebar?.title) window.titlebar.title.innerHTML = title;
  }
};

export const registerWindowUpdateTitle = (): void => {
  TopBarController.onTitleChange(updateTitle);
};

export const unregisterWindowUpdateTitle = (): void => {
  TopBarController.offTitleChange(updateTitle);
};

export default FileName;
