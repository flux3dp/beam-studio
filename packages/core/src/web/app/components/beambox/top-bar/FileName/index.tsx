import React, { useEffect } from 'react';

import TopBarIcons from '@core/app/icons/top-bar/TopBarIcons';
import currentFileManager from '@core/app/svgedit/currentFileManager';
import TopBarController from '@core/app/views/beambox/TopBar/contexts/TopBarController';
import useForceUpdate from '@core/helpers/use-force-update';
import useI18n from '@core/helpers/useI18n';

import styles from './index.module.scss';

interface Props {
  hasUnsavedChange: boolean;
  isTitle?: boolean;
}

function FileName({ hasUnsavedChange, isTitle = false }: Props): React.JSX.Element {
  const forceUpdate = useForceUpdate();

  useEffect(() => {
    TopBarController.onTitleChange(forceUpdate);

    return () => {
      TopBarController.offTitleChange(forceUpdate);
    };
  }, [forceUpdate]);

  const lang = useI18n().topbar;
  const { isCloudFile } = currentFileManager;
  const fileName = currentFileManager.getName() || lang.untitled;

  return (
    <div className={isTitle ? styles.title : styles['file-name']}>
      {isCloudFile && <TopBarIcons.CloudFile className={styles.cloud} />}
      {`${fileName}${hasUnsavedChange ? '*' : ''}`}
    </div>
  );
}

export default FileName;
