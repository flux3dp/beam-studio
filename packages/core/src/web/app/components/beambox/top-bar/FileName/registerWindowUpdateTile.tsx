import ReactDomServer from 'react-dom/server';

import TopBarController from '@core/app/views/beambox/TopBar/contexts/TopBarController';

import FileName from '.';

const updateTitle = () => {
  if (window.os === 'Windows' && window.titlebar) {
    const title = ReactDomServer.renderToStaticMarkup(<FileName hasUnsavedChange={false} isTitle />);

    if (window.titlebar?.title) {
      window.titlebar.title.innerHTML = title;
    }
  }
};

const unregisterWindowUpdateTitle = (): void => {
  TopBarController.offTitleChange(updateTitle);
};

export const registerWindowUpdateTitle = (): (() => void) => {
  TopBarController.onTitleChange(updateTitle);

  return unregisterWindowUpdateTitle;
};
