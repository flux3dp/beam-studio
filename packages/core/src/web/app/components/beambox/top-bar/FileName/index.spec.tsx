import React from 'react';

import { render } from '@testing-library/react';

import FileName from './index';

const mockGetName = jest.fn();
const mockGetIsCloudFile = jest.fn();

jest.mock('@core/app/svgedit/currentFileManager', () => ({
  __esModule: true,
  default: {
    getName: () => mockGetName(),
    get isCloudFile() {
      return mockGetIsCloudFile();
    },
  },
}));

const mockOnTitleChange = jest.fn();
const mockOffTitleChange = jest.fn();

jest.mock('@core/app/views/beambox/TopBar/contexts/TopBarController', () => ({
  offTitleChange: (...args) => mockOffTitleChange(...args),
  onTitleChange: (...args) => mockOnTitleChange(...args),
}));

describe('test FileName', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    mockGetIsCloudFile.mockReturnValue(false);
    mockGetName.mockReturnValue('abc');

    const { container, rerender } = render(<FileName hasUnsavedChange />);

    expect(container).toMatchSnapshot();

    mockGetName.mockReturnValue(null);
    rerender(<FileName hasUnsavedChange={false} />);
    expect(container).toMatchSnapshot();

    mockGetName.mockReturnValue(null);
    rerender(<FileName hasUnsavedChange={false} isTitle />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with cloud file', () => {
    mockGetIsCloudFile.mockReturnValue(true);
    mockGetName.mockReturnValue('abc');

    const { container, rerender } = render(<FileName hasUnsavedChange />);

    expect(container).toMatchSnapshot();

    mockGetName.mockReturnValue(null);
    rerender(<FileName hasUnsavedChange={false} />);
    expect(container).toMatchSnapshot();

    mockGetName.mockReturnValue(null);
    rerender(<FileName hasUnsavedChange={false} isTitle />);
    expect(container).toMatchSnapshot();
  });
});
