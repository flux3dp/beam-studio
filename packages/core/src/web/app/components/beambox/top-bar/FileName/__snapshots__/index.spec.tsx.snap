// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test FileName should render correctly 1`] = `
<div>
  <div
    class="file-name"
  >
    abc*
  </div>
</div>
`;

exports[`test FileName should render correctly 2`] = `
<div>
  <div
    class="file-name"
  >
    Untitled
  </div>
</div>
`;

exports[`test FileName should render correctly 3`] = `
<div>
  <div
    class="title"
  >
    Untitled
  </div>
</div>
`;

exports[`test FileName should render correctly with cloud file 1`] = `
<div>
  <div
    class="file-name"
  >
    <svgr-url
      classname="cloud"
    />
    abc*
  </div>
</div>
`;

exports[`test FileName should render correctly with cloud file 2`] = `
<div>
  <div
    class="file-name"
  >
    <svgr-url
      classname="cloud"
    />
    Untitled
  </div>
</div>
`;

exports[`test FileName should render correctly with cloud file 3`] = `
<div>
  <div
    class="title"
  >
    <svgr-url
      classname="cloud"
    />
    Untitled
  </div>
</div>
`;
